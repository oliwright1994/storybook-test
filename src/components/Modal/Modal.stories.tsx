import createStory from '@bright/storybook/createStory'
import { Box } from '@chakra-ui/core'
import { boolean, CHANGE, text } from '@storybook/addon-knobs'
import { addons } from '@storybook/addons'
import React from 'react'

import Button from '../Button'
import Input from '../Input'
import Text from '../Text'
import Modal from '.'

createStory('COMPONENTS | Modal', [
  {
    label: 'Modal',
    jsx: () => {
      const channel = addons.getChannel()
      const isOpenIdentifier = 'Modal open'
      const onOpen = (): void => {
        channel.emit(CHANGE, {
          name: isOpenIdentifier,
          value: true,
        })
      }

      const onClose = (): void => {
        channel.emit(CHANGE, {
          name: isOpenIdentifier,
          value: false,
        })
      }

      const isOpen = boolean(isOpenIdentifier, false)
      const title = text('Modal title', 'The Test Modal')
      return (
        <>
          <Button onClick={onOpen}>Open Modal</Button>
          <Modal isOpen={isOpen} onClose={onClose}>
            <Modal.Header>{title}</Modal.Header>
            <Modal.CloseButton />
            <Modal.Body>
              <Text>
                Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et
                dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip
                ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu
                fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia
                deserunt mollit anim id est laborum.
              </Text>
              <Box p="5">
                <Input
                  id="confirm"
                  onChange={() => {}}
                  label="Confirm"
                  description={'Please type "CONFIRM" below to confirm your submission'}
                />
              </Box>
            </Modal.Body>
            <Modal.Footer>
              <Button onClick={() => {}}> Submit</Button>
            </Modal.Footer>
          </Modal>
        </>
      )
    },
  },
])
