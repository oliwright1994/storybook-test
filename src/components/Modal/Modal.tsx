import Body from '@bright/components/Modal/Body'
import CloseButton from '@bright/components/Modal/CloseButton'
import Footer from '@bright/components/Modal/Footer'
import Header from '@bright/components/Modal/Header'
import {
  IModal as IChakraModal,
  Modal as ChakraModal,
  ModalContent as ChakraModalContent,
  ModalOverlay as ChakraModalOverlay,
} from '@chakra-ui/core'
import React, { FC } from 'react'

type IModalProps = Pick<IChakraModal, 'initialFocusRef' | 'returnFocusOnClose' | 'isOpen' | 'onClose'>

interface IModal extends FC<IModalProps> {
  CloseButton: typeof CloseButton
  Header: typeof Header
  Body: typeof Body
  Footer: typeof Footer
}

const Modal: IModal = ({ children, isOpen, onClose }) => {
  return (
    <ChakraModal
      addAriaLabels
      blockScrollOnMount
      closeOnOverlayClick
      closeOnEsc
      scrollBehavior="inside"
      size="md"
      returnFocusOnClose
      useInert
      isOpen={isOpen}
      onClose={onClose}
    >
      <ChakraModalOverlay backgroundColor="blackAlpha.400" />
      <ChakraModalContent>{children}</ChakraModalContent>
    </ChakraModal>
  )
}

Modal.CloseButton = CloseButton
Modal.Header = Header
Modal.Body = Body
Modal.Footer = Footer

export default Modal
