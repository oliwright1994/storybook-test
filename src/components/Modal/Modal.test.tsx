import { act, fireEvent } from '@testing-library/react'
import React from 'react'
import { renderWithBrightTheme } from 'testing/helpers/themeHelper'

import Modal from '.'

describe('<Modal />', () => {
  const defaultText = 'Test Text'

  it('should render as expected', () => {
    const { container } = renderWithBrightTheme(<Modal />)

    expect(container).toMatchSnapshot()
  })

  it('should only be present when isOpen is true', () => {
    let isOpen = false
    const { queryByText, rerender } = renderWithBrightTheme(<Modal isOpen={isOpen}>{defaultText}</Modal>)

    expect(queryByText(defaultText)).toBeNull()

    isOpen = true
    rerender(<Modal isOpen={isOpen}>{defaultText}</Modal>)

    expect(queryByText(defaultText)).toBeTruthy()
  })

  it('should fire onClose handler when closeButton is clicked', () => {
    const onClose = jest.fn()
    const { getByLabelText } = renderWithBrightTheme(
      <Modal isOpen onClose={onClose}>
        <Modal.CloseButton />
        {defaultText}
      </Modal>
    )

    const closeBtn = getByLabelText('Close')
    expect(onClose).toHaveBeenCalledTimes(0)

    act(() => {
      fireEvent.click(closeBtn)
    })

    expect(onClose).toHaveBeenCalledTimes(1)
  })

  it('should render namespaced subcomponents correctly', () => {
    const { getAllByText, getByLabelText } = renderWithBrightTheme(
      <Modal isOpen>
        <Modal.CloseButton />
        <Modal.Header>{defaultText}</Modal.Header>
        <Modal.Body>{defaultText}</Modal.Body>
        <Modal.Footer>{defaultText}</Modal.Footer>
      </Modal>
    )

    const closeBtnEl = getByLabelText(/close/i)
    expect(closeBtnEl).toBeTruthy()

    const sectionElements = getAllByText(defaultText)
    expect(sectionElements).toHaveLength(3)
  })

  it('should have proper "aria" attributes', () => {
    const { getByRole, getByText } = renderWithBrightTheme(
      <Modal isOpen>
        <Modal.CloseButton />
        <Modal.Header>Modal Header</Modal.Header>
        <Modal.Body>Modal body</Modal.Body>
        <Modal.Footer>{defaultText}</Modal.Footer>
      </Modal>
    )

    const dialog = getByRole('dialog')

    expect(dialog).toHaveAttribute('aria-modal', 'true')

    expect(getByText(/modal header/i).id).toEqual(dialog.getAttribute('aria-labelledby'))

    expect(getByText(/modal body/i).id).toEqual(dialog.getAttribute('aria-describedby'))
  })
})
