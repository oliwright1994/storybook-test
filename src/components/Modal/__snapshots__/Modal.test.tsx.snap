// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render as expected 1`] = `<div />`;
